File: web/backend/src/models/match.ts
import {Request} from 'express';
import {MongoDb} from '../db';
import {DataValidDTO} from './DTOs/dataValidDTO';

export class Match {

    public static async validate(req: Request) {
        if (!Object.keys(req.body).length) {
            return new DataValidDTO(false, 'A team object is required in the body of this request');
        }

        if (!req.body.League) {
            return new DataValidDTO(false, 'A league must be specified for this match');
        }

        if (!req.body.Home || req.body.Home.length !== MongoDb.MONGO_ID_LEN) {
            return new DataValidDTO(false, 'A home team must be specified');
        }

        if (!req.body.Away || req.body.Away.length !== MongoDb.MONGO_ID_LEN) {
            return new DataValidDTO(false, 'A home away must be specified');
        }
        if (req.body.Away === req.body.Home) {
            return new DataValidDTO(false, 'A team cannot play itself');
        }
        const homeTeam = await MongoDb.getById('team', req.body.Home);
        const awayTeam = await MongoDb.getById('team', req.body.Away);

        if (!homeTeam.valid || !homeTeam.data) {
            return new DataValidDTO(false, 'The specified home team was not found');
        }
        if (!awayTeam.valid || !awayTeam.data) {
            return new DataValidDTO(false, 'The specified away team was not found');
        }
        return new DataValidDTO(true, '');
    }

    public static async validateUpdate(req: Request) {
        if (!Object.keys(req.body).length) {
            return new DataValidDTO(false, 'A team object is required in the body of this request');
        }

        if (!req.body.Home) {
            if (req.body.Home.length < 1) {
                return new DataValidDTO(false, 'A home team must be specified');
            }
        }
        if (!req.body.Away) {
            if (req.body.Away.length < 1) {
                return new DataValidDTO(false, 'An away team must be specified');
            }
        }
    }

    public _id: string; // The match id
    public Title: string;
    public Home: string; // The id of the home team
    public Away: string; // The id of the away team
    public Victor: string; // The id of the winning team
    public League: string; // The id of the league the match is in
    public Loser: string; // The id of the losing team
    public Home_Score: number = 0;
    public Away_Score: number = 0;
    public Fill_From: [number, number];
    public Tournament: string; // The id of the tournament the match is in. Not required
    public Updated_By: string;
    public Status: Match_Status;

    constructor(req, homeTeam, awayTeam) {
        this.Home = req.Home;
        this.Away = req.Away;
        this.Title = homeTeam.Name + ' VS ' + awayTeam.Name;
        this.League = req.League;
        this.Status = Match_Status.In_progress;
        if (req.Tournament) {
            this.Tournament = req.Tournament;
        } else {
            this.Tournament = 'Not part of a Tournament';
        }
    }
}

export enum Match_Status {
    In_progress,
    Pending_report,
    Confirmed,
    Conflicted,
}

File: web/v2/backend/src/main/java/com/tourneygen/web/Models/Match.java
package com.tourneygen.web.Models;

import com.tourneygen.web.Models.DTOs.MatchDTO;
import com.tourneygen.web.Models.Repositories.LeagueRepository;
import com.tourneygen.web.Models.Repositories.TeamRepository;
import java.util.Date;
import javax.persistence.*;

@Entity
@Table(name = "Matches")
public class Match {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @ManyToOne @JoinColumn private Team homeTeam;

  @ManyToOne @JoinColumn private Team awayTeam;

  private String title;

  private Date date;

  private @ManyToOne @JoinColumn Team victor;

  private @ManyToOne @JoinColumn Team loser;

  private @ManyToOne @JoinColumn League league;

  private int homeScore = 0;

  private int awayScore = 0;

  // TODO Add tournament field

  private Long updatedBy;

  private String status = "In_Progress";

  public Match() {}

  public Long getId() {
    return id;
  }

  public void setId(Long id) {
    this.id = id;
  }

  public Team getHomeTeam() {
    return homeTeam;
  }

  public void setHomeTeam(Team homeTeam) {
    this.homeTeam = homeTeam;
  }

  public Team getAwayTeam() {
    return awayTeam;
  }

  public void setAwayTeam(Team awayTeam) {
    this.awayTeam = awayTeam;
  }

  public String getTitle() {
    return title;
  }

  public void setTitle(String title) {
    this.title = title;
  }

  public Date getDate() {
    return date;
  }

  public void setDate(Date date) {
    this.date = date;
  }

  public Team getVictor() {
    return victor;
  }

  public void setVictor(Team victor) {
    this.victor = victor;
  }

  public Team getLoser() {
    return loser;
  }

  public void setLoser(Team loser) {
    this.loser = loser;
  }

  public League getLeague() {
    return league;
  }

  public void setLeague(League league) {
    this.league = league;
  }

  public int getHomeScore() {
    return homeScore;
  }

  public void setHomeScore(int homeScore) {
    this.homeScore = homeScore;
  }

  public int getAwayScore() {
    return awayScore;
  }

  public void setAwayScore(int awayScore) {
    this.awayScore = awayScore;
  }

  public Long getUpdatedBy() {
    return updatedBy;
  }

  public void setUpdatedBy(Long updatedBy) {
    this.updatedBy = updatedBy;
  }

  public String getStatus() {
    return status;
  }

  public void setStatus(String status) {
    this.status = status;
  }

  public void create(
      MatchDTO matchDTO, LeagueRepository leagueRepository, TeamRepository teamRepository) {
    this.title = matchDTO.getTitle();
    this.date = matchDTO.getDate();
    this.homeTeam =
        teamRepository
            .findById(matchDTO.getHomeId())
            .orElseThrow(
                () -> new EntityNotFoundException("Home Team with id " + id + " was not found"));
    this.awayTeam =
        teamRepository
            .findById(matchDTO.getAwayId())
            .orElseThrow(
                () -> new EntityNotFoundException("Away Team with id " + id + " was not found"));
    this.league =
        leagueRepository
            .findById(matchDTO.getLeagueId())
            .orElseThrow(
                () -> new EntityNotFoundException("League with id " + id + " was not found"));
  }
}
